ip-utils.c

```diff
@@ -12,13 +12,45 @@
  */
 
 #include <errno.h>
+#include <sys/socket.h>
 #include <sys/types.h>
 #include <arpa/inet.h>
+#include <netdb.h>
 #include <net/if.h>
 #include <string.h>
+#include <unistd.h>
+
+#include <glib.h>
 
 #include "ip-utils.h"
 
+int janus_network_detect_local_ip(char *buf, size_t buflen) {
+	struct sockaddr_in addr;
+	socklen_t len;
+	int fd = socket(AF_INET, SOCK_DGRAM, 0);
+	if(fd == -1)
+		goto error;
+	addr.sin_family = AF_INET;
+	addr.sin_port = htons(1);
+	inet_pton(AF_INET, "1.2.3.4", &addr.sin_addr.s_addr);
+	if(connect(fd, (const struct sockaddr*) &addr, sizeof(addr)) < 0)
```

Comment:
The only purpose of this method is not just to find a local IP address, but to find an IP address that is likely to be used for communicating with the outside. We create a file descriptor and we point it to an external address: the UDP connect doesn't really connect, but still involves the OS routing table to see which interface would be able to get there. We then get the local address from the socket.

I think this was contributed a long time ago in a PR, and seems to work fine. But yes, it's currently limited to IPv4 and should be improved to support IPv6 too.